testing/travis.py

```diff
@@ -0,0 +1,248 @@
+    commit_range = os.environ.get("TRAVIS_COMMIT_RANGE", "master..HEAD")
+    diff_names = str(subprocess.check_output(["git",  "diff",  "--name-only", commit_range]))
+
+    not_docfiles = filter_not_docfiles(diff_names)
+    if len(not_docfiles) < 1:
+        logging.info("All of the changes were in documentation. Skipping build.")
+        sys.exit(0)
+
+    linting_providers = ["clc", "softlayer", "triton"]
+    deploy_providers = ["gce", "aws", "do"]
+
+    cmds = linter_cmds()
+
+    if os.environ.get("PROVIDER", None) in deploy_providers:
+        cmds.extend(deploy_to_cloud())
+
+    if not run_cmds(cmds, fail_sequential=True):
+        sys.exit(1)
+
+    health_checks()
+
+    sys.exit(0)
+
+
+def after_script():
+    """Cleanup after ci_build"""
+
+    destroy_cmd = ["terraform",  "destroy",  "-force"]
+    logging.info("Destroying cloud provider resources")
+
+    sys.exit(run_cmd(destroy_cmd))
+    # send slack notification
+
+
+## PURE-ISH FUNCTIONS
+
+def run_cmd(cmd, attempts=1):
+    """ Runs a command attempts times, logging its output. Returns True if it
+    succeeds once, or False if it never does. """
+    try:
+        for i in range(attempts):
+            proc = subprocess.Popen(cmd, stdin=open(os.devnull, "r"))
+            proc.wait()
+            if not proc.returncode == 0:
+                print("Command ", str(cmd), " failed")
+            else:
+                print("Command ", str(cmd), " succeeded.")
+                return True
+            time.sleep(3)
+        return False
+    except OSError as e:
+        print("Error while attempting to run command ", cmd)
+        print(e)
+
+
+def run_cmds(cmds, fail_sequential=False):
+    """ Run a list of ([args], tries) tuples, aborting the run if a single one
+    fails and fail_sequential is set to True.
+    Returns True if all succeed, and False if any fail.
+    """
+    to_return = True
+    for (cmd, tries) in cmds:
+        # If we've failed already and fail_sequential is set, don't continue
+        if not to_return and fail_sequential:
+            return False
+        # Otherwise, attempt to run the current command
+        elif run_cmd(cmd, attempts=tries):
+            continue
+        # If it fails, set the exit code appropriately
+        else:
+            to_return = False
+    return to_return
+
+
+def filter_not_docfiles(diff_names):
+
+    not_docfiles = []
+    for diff_name in diff_names.split():
+        if diff_name.startswith('docs'):
+            continue
+        elif diff_name.endswith('md'):
+            continue
+        elif diff_name.endswith('rst'):
+            continue
+        else:
+            not_docfiles.append(diff_name)
+
+    return not_docfiles
+
+
+def linter_cmds():
+    return [
+        (["ssh-add"], 1),
+        (["terraform", "get"], 1),
+        (["terraform", "plan", "--input=false", "--refresh=false"], 1),
+    ]
+
+
+def deploy_to_cloud_cmds():
+    ap = [
```

Comment:
This could be named clearer, like `ansible_playbook` or something. I think that one's my bad :)